packages/synapse-react-client/src/components/TableColumnSchemaEditor/Validators/ColumnModelValidator.ts

```diff
@@ -0,0 +1,145 @@
+ */
+export const columnModelZodSchema = columnModelBaseZodSchema
+  .merge(
+    z.object({
+      id: z.string().optional(),
+      defaultValue: z.union([z.string(), z.array(z.any())]).optional(),
+      maximumSize: z
+        .union([optionalStringSchema, z.number()])
+        .pipe(
+          z.coerce
+            .number()
+            .finite()
+            .int()
+            .min(1)
+            .max(MAX_STRING_SIZE)
+            .optional(),
+        ),
+      maximumListLength: z
+        .union([optionalStringSchema, z.number()])
+        .pipe(
+          z.coerce
+            .number()
+            .finite()
+            .int()
+            .min(1)
+            .max(MAX_LIST_LENGTH)
+            .optional(),
+        ),
+      enumValues: z.array(z.coerce.string()).optional(),
+      jsonSubColumns: z.array(jsonSubColumnModelZodSchema).optional(),
+      facetType: facetTypeSchema.optional(),
+    }),
+  )
+  .refine(data => data.maximumSize == null || canHaveSize(data.columnType), {
+    message: 'Size is not allowed for this column type',
+    path: ['maximumSize'],
+  })
+  .refine(
+    data =>
+      data.maximumListLength == null || canHaveMaxListLength(data.columnType),
+    {
+      message: 'Maximum list length is not allowed for this column type',
+      path: ['maximumListLength'],
+    },
+  )
+  .refine(
+    data => {
+      if (data.jsonSubColumns != null) {
+        return data.columnType === ColumnTypeEnum.JSON
+      }
+      return true
+    },
+    {
+      message: 'Only JSON columns can have JSON sub-columns.',
+      path: ['columnType'],
+    },
+  )
+  .transform((data, ctx) => {
+    if (data.defaultValue != null) {
+      // Validate and transform the defaultValue based on the columnType
+      const defaultValueSchema = getDefaultValueValidator(data.columnType)
+      const result = defaultValueSchema.safeParse(data.defaultValue)
+      let transformedValue: string | undefined
+      if (result.success) {
+        transformedValue = result.data
+      } else {
+        result.error.issues.forEach(issue => {
+          ctx.addIssue({ ...issue, path: ['defaultValue'] })
+        })
+        transformedValue = undefined
+      }
+      return { ...data, defaultValue: transformedValue }
+    }
+    return omit(data, 'defaultValue')
+  })
+  .transform((data, ctx) => {
+    if (data.enumValues != null) {
+      // Validate and transform the defaultValue based on the columnType
+      const enumValuesSchema = getEnumValuesValidator(data.columnType)
+      const result = enumValuesSchema.safeParse(data.enumValues)
+      let transformedValue: string[] | undefined
+      if (result.success) {
+        transformedValue = result.data
+      } else {
+        result.error.issues.forEach(issue => {
+          ctx.addIssue({ ...issue, path: ['enumValues'] })
+        })
+        transformedValue = undefined
+      }
+      return { ...data, enumValues: transformedValue }
+    }
+    return omit(data, 'enumValues')
```

Comment:
This is the exact same set of steps as the previous `transform`, but on a different field, and different data validator. Do you think this is worth extracting to a new function that takes (fieldName, validatorProvider)?

Reply:
> Do you think this is worth extracting to a new function that takes (fieldName, validatorProvider)?

I think this slight refactor is worth doing

Reply:
I spent some time trying to do this and it's very hard, if not impossible, to write a function that doesn't break Zod's ability to infer the output type of the parser. I will not refactor this at this time, but I will add a comment to document this.